main.c

```diff
@@ -486,22 +430,76 @@ int main(void) {
   halInit();
   chSysInit();
 
-  profiles[currentProfile].callback(ledColors, ledIntensity);
+  /* profiles[currentProfile].callback(ledColors, ledIntensity); */
   updateAnimationSpeed();
 
   // Setup masks to all be 0xFF at the start
   memset(ledMasks, 0xFF, sizeof(ledMasks));
 
+  chBSemObjectInit(&ledDisabledSem, true);
+
   palClearLine(LINE_LED_PWR);
   sdStart(&SD1, &usart1Config);
 
-  // Setup Column Multiplex Timer
-  gptStart(&GPTD_BFTM0, &bftm0Config);
-  gptStartContinuous(&GPTD_BFTM0, SERIAL_CHECK_FREQUENCY);
+  chThdCreateStatic(waThread1, sizeof(waThread1), HIGHPRIO, MsgHandlerThd,
+                    NULL);
 
-  // idle loop, copied form lib/chibios/os/rt/src/chsys.c
   while (true) {
-    port_wait_for_interrupt();
-    CH_CFG_IDLE_LOOP_HOOK();
+
+    // If leds are disabled or enabled and using a reactive profile which does
+    // not need any update, suspend the thread on the semaphore. The semaphore
+    // is flagged up in MsgHandlerThd when leds are enabled or user clicks any
+    // button with the reactive profile enabled.
+    if (!ledEnabled && !reactiveNeedsUpdate) {
+      chBSemWait(&ledDisabledSem);
+    } else {
+      if (enableRow) {
+        palClearLine(ledRows[currentRow]);
+      }
+
+      enableRow = false;
+      currentRow = (currentRow + 1) % (NUM_ROW * 4);
+      if (currentRow % 4 == 3) {
+        currentRow = (currentRow + 1) % (NUM_ROW * 4);
+      }
+
+      rowPWMCount += 63;
```

Comment:
This is a bit weird hack. Let's say RED color intensity is 5. It means RED is enabled for 5 rows and disabled for 251 if we use sequential `rowPWMCount`. By using `63` increment I sort of randomize the distribution of `rowPWMCount`. The number is chosen such that for 8 bit unsigned value with wrap-around we don't repeat any number until we exhaust all 256 values.

Reply:
Please document this in comments. As this might be useful to later people who reads the code

Reply:
Added the comment.